aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -683,6 +683,7 @@ Tensor randperm(int64_t n, const TensorOptions& options) {
 
 Tensor randperm(int64_t n, c10::optional<Generator> generator, const TensorOptions& options) {
   auto tensor = at::empty(n, options);
+  TORCH_CHECK(!generator.has_value() || (generator.has_value() && tensor.device().type() == generator->device().type()), "Expected a '", tensor.device().type(), "' generator device but found '", generator->device().type(), "'");
```

Comment:
Generators have a full device associated with them, so it may be better compare `device()` rather than `device().type()`.

However, it is true that the device index on CUDA generators doesn't actually matter, as no RNG state is actually stored on GPU. Still, probably better to enforce this invariant to give us more options if we want to change how the internal implementation works.